### PR TITLE
Tor

### PR DESCRIPTION
**Pull request by bonomat**
This PR adds a small lib which allows us to add a Tor service easily. The whole thing is hidden behind a feature-flag so that it does not bother us. 

Note: GH action might change if it fails, the rest is ready for review.

Kudos to @luckysori for helping with the annoying handler type.

### CODE REVIEW

File: xmr-btc/src/tor.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    control::{AsyncEvent, AuthenticatedConn, ConnError, UnauthenticatedConn},
+    onion::TorSecretKeyV3,
+};
+
+lazy_static! {
+    /// The default TOR socks5 proxy address, `127.0.0.1:9050`.
+    pub static ref TOR_PROXY_ADDR: SocketAddrV4 = SocketAddrV4::new(Ipv4Addr::LOCALHOST, 9050);
+    /// The default TOR Controller Protocol address, `127.0.0.1:9051`.
+    pub static ref TOR_CP_ADDR: SocketAddr = SocketAddr::V4(SocketAddrV4::new(Ipv4Addr::LOCALHOST, 9051));
+}
```
Comment (luckysori) [inline xmr-btc/src/tor.rs:18]: :wrench: These are only ever used as `&str`. Could we initialise them as such using `const`s without needing `lazy_static`? Or is this providing something more?

File: xmr-btc/tests/tor.rs [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+    #[tokio::test]
+    async fn test_tor_control_port() -> Result<()> {
+        // Setup test HTTP Server
+        let (tx, rx) = tokio::sync::oneshot::channel::<()>();
+        let port = 8080;
+        start_test_service(port, rx);
+
+        // Connect to local Tor service
+        let mut authenticated_connection = AuthenticatedConnection::new().await?;
```
Comment (luckysori) [inline xmr-btc/tests/tor.rs:43]: :thinking: I guess this means that we assume that Tor is running before this test is run. We should skip this in CI and document it if that's the case.

**Review by luckysori — COMMENTED** [review]
Just a couple of things I noticed.

File: xmr-btc/src/tor.rs (hunk 1 above)
Comment (bonomat) [reply xmr-btc/src/tor.rs]: Removed in the last patch.

File: xmr-btc/tests/tor.rs (hunk 2 above)
Comment (bonomat) [reply xmr-btc/tests/tor.rs]: I'm starting Tor from the test now. For that it assumes that Tor is in the path. 
Not sure if this is the nicest way to go. Maybe mentioning it in the readme is enough? 
It's behind a feature flag anyway.

Comment (bonomat) [reply xmr-btc/tests/tor.rs]: I'm starting Tor from the test now. For that it assumes that Tor is in the path. 
Not sure if this is the nicest way to go. Maybe mentioning it in the readme is enough? 
It's behind a feature flag anyway.

File: xmr-btc/src/tor.rs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+                control_port,
+            )),
+        }
+    }
+
+    /// checks if tor is running
+    async fn tor_running(&self) -> Result<()> {
+        // Make sure you are running tor and this is your socks port
+        let proxy = reqwest::Proxy::all(format!("socks5h://{}", self.tor_proxy_address).as_str())
+            .expect("tor proxy should be there");
```
Comment (tcharding) [inline xmr-btc/src/tor.rs:42]: Can we return an error here instead of panicing please.

File: xmr-btc/src/tor.rs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            tor_proxy_address: SocketAddrV4::new(Ipv4Addr::LOCALHOST, proxy_port),
+            tor_control_port_address: SocketAddr::V4(SocketAddrV4::new(
+                Ipv4Addr::LOCALHOST,
+                control_port,
+            )),
+        }
+    }
+
+    /// checks if tor is running
+    async fn tor_running(&self) -> Result<()> {
```
Comment (tcharding) [inline xmr-btc/src/tor.rs:39]: Since this returns an error (i.e., this function is not a predicate), perhaps we could call it `assert_tor_running`

File: xmr-btc/src/tor.rs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        let res = client.get("https://check.torproject.org").send().await?;
+
+        let text = res.text().await?;
+        let is_tor = text.contains("Congratulations. This browser is configured to use Tor.");
+
+        if is_tor {
+            Ok(())
+        } else {
+            bail!("Tor is currently not running")
+        }
```
Comment (tcharding) [inline xmr-btc/src/tor.rs:57]: Perhaps
```
        if !text.contains("Congratulations. This browser is configured to use Tor.") {
            bail!("Tor is currently not running")
        }

        Ok(())
```

File: xmr-btc/src/tor.rs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    }
+
+    async fn init_unauthenticated_connection(&self) -> Result<UnauthenticatedConn<TcpStream>> {
+        // Connect to local tor service via control port
+        let sock = TcpStream::connect(self.tor_control_port_address).await?;
+        let unauthenticated_connection = UnauthenticatedConn::new(sock);
+        Ok(unauthenticated_connection)
```
Comment (tcharding) [inline xmr-btc/src/tor.rs:64]: This is a local variable with very limited scope, such a long identifier does not add any additional information. Perhaps we could use
```
let uc = UnauthenticatedConn::new(sock);
Ok(uc)
```

File: xmr-btc/src/tor.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+        let mut unauthenticated_connection = match self.init_unauthenticated_connection().await {
+            Err(_) => bail!("Tor instance not running"),
+            Ok(unauthenticated_connection) => unauthenticated_connection,
+        };
+
+        let tor_info = match unauthenticated_connection.load_protocol_info().await {
+            Ok(info) => info,
+            Err(_) => bail!("Failed to load protocol info from Tor."),
+        };
```
Comment (tcharding) [inline xmr-btc/src/tor.rs:79]: We can use `map_err` here
```
        let tor_info = unauthenticated_connection
            .load_protocol_info()
            .await
            .map_err(|_| anyhow!("Failed to load protocol info from Tor."))?;
```

File: xmr-btc/src/tor.rs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+                None,
+                &mut [(
+                    port,
+                    SocketAddr::new(IpAddr::from(Ipv4Addr::new(127, 0, 0, 1)), port),
+                )]
+                .iter(),
+            )
+            .await
+            .map_err(|_| anyhow!("Could not add onion service."))?;
+        Ok(())
```
Comment (tcharding) [inline xmr-btc/src/tor.rs:125]: We can just return this directly, no need to use `?;` and then `Ok(())`

**Review by tcharding — APPROVED** [review]
Seems functional. This `torut` library makes it hard to write clean code. ACK'ing as is but please consider two points
1. The types `AuthenticatedConnection` and `UnauthenticatedConnection` are not uniform (not to mention `AuthenticatedConn<TcpStream>`, it would be nice if it was all a bit more balanced. What I mean is unauth just has sockets in it and has conversion methods but auth is totally different - this is unexpected but may not be fixable due to the `torut` library.
2. Local variables are extremely long, please consider using `ac` and `uc` instead of `authenticated_connection`.

File: xmr-btc/src/tor.rs (hunk 3 above)
Comment (bonomat) [reply xmr-btc/src/tor.rs:42]: Good point

File: xmr-btc/src/tor.rs (hunk 5 above)
Comment (bonomat) [reply xmr-btc/src/tor.rs:57]: Looks good, will do that.

File: xmr-btc/src/tor.rs (hunk 7 above)
Comment (bonomat) [reply xmr-btc/src/tor.rs:79]: Is this to be preferred over `bail!` in general?

Comment (tcharding) [reply xmr-btc/src/tor.rs:79]: Its functionally exactly the same, just more concise (IMO).

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -77,6 +83,7 @@ jobs:
         run: cargo check --workspace --all-targets --all-features
 
       - name: Cargo test
-        run: cargo test --workspace --all-features
+        run: cargo test -p xmr-btc test_tor_control_port --all-features
```
Comment (tcharding) [inline .github/workflows/ci.yml:86]: Why this change please?

**Review by tcharding — APPROVED** [review]
LGTM, except the test change CI question.

File: .github/workflows/ci.yml (hunk 9 above)
Comment (bonomat) [reply .github/workflows/ci.yml]: To get the tests running :D Obviously this should not have been in this PR.

File: .github/workflows/ci.yml [hunk 10]
```diff
@@ -46,9 +46,20 @@ jobs:
...
         uses: actions/checkout@v2
 
+      - name: Install and stop tor in case it was running
+        run: |
+          sudo apt install software-properties-common
+          sudo curl https://deb.torproject.org/torproject.org/A3C4F0F979CAA22CDBA8F512EE8CBC9E886DDD89.asc | sudo gpg --import
+          sudo gpg --export A3C4F0F979CAA22CDBA8F512EE8CBC9E886DDD89 | sudo apt-key add -
+          sudo add-apt-repository 'deb https://deb.torproject.org/torproject.org bionic main'
+          sudo apt update
+          sudo apt install tor deb.torproject.org-keyring
```
Comment (bonomat) [inline .github/workflows/ci.yml:60]: @tcharding : Is this the way how you installed Tor on your machine btw? 
My issue with CI came from an outdated Tor version

File: xmr-btc/tests/tor.rs [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+    fn start_test_service(port: u16, rx: Receiver<()>) {
+        let make_svc =
+            make_service_fn(|_conn| async { Ok::<_, Infallible>(service_fn(hello_world)) });
+        let addr = ([127, 0, 0, 1], port).into();
+        let server = hyper::Server::bind(&addr).serve(make_svc);
+        let graceful = server.with_graceful_shutdown(async {
+            rx.await.ok();
+        });
+        tokio::spawn(async {
+            // server.await.unwrap();
```
Comment (bonomat) [inline xmr-btc/tests/tor.rs:33]: !
